File: src/eQuantic.UI.Runtime/src/utils/class-builder.ts
/**
 * ClassBuilder - Fluent builder for combining CSS classes
 *
 * TypeScript implementation that mirrors eQuantic.UI.Core.Styling.ClassBuilder (C#).
 * Provides a clean, chainable API for building CSS class strings with support for:
 * - Conditional classes
 * - State variants (hover, focus, active, dark mode)
 * - Responsive breakpoints
 * - Automatic duplicate removal
 */

export class ClassBuilder {
  private classes: string[] = [];
  private added: Set<string> = new Set();

  /**
   * Creates a new empty ClassBuilder instance.
   */
  static create(...initialClasses: (string | null | undefined)[]): ClassBuilder {
    const builder = new ClassBuilder();
    if (initialClasses.length > 0) {
      builder.add(...initialClasses);
    }
    return builder;
  }

  /**
   * Adds one or more CSS classes.
   * Duplicate classes are automatically ignored.
   */
  add(...classes: (string | null | undefined)[]): this {
    for (const className of classes) {
      if (className && className.trim() && !this.added.has(className)) {
        this.added.add(className);
        this.classes.push(className);
      }
    }
    return this;
  }

  /**
   * Conditionally adds classes if the condition is true.
   * Supports optional fallback classes when condition is false.
   *
   * @example
   * .when(isActive, "bg-blue-600", "bg-gray-600")
   * .when(isDisabled, "opacity-50")
   */
  when(condition: boolean, ...classes: (string | null | undefined)[]): this;
  when(condition: boolean, whenTrue: string, whenFalse?: string): this;
  when(condition: boolean, ...args: any[]): this {
    if (condition) {
      // If first arg is string and there are exactly 1-2 args, treat as whenTrue
      if (typeof args[0] === 'string' && args.length <= 2) {
        this.add(args[0]);
      } else {
        // Otherwise, treat as multiple classes
        this.add(...args);
      }
    } else if (args.length === 2 && typeof args[1] === 'string') {
      // Second argument is the fallback (whenFalse)
      this.add(args[1]);
    }
    return this;
  }

  /**
   * Adds classes with a custom prefix.
   *
   * @example
   * .withPrefix("hover:", "bg-gray-100", "scale-110")
   * // => "hover:bg-gray-100 hover:scale-110"
   */
  withPrefix(prefix: string, ...classes: (string | null | undefined)[]): this {
    for (const className of classes) {
      if (className && className.trim()) {
        this.add(`${prefix}${className}`);
      }
    }
    return this;
  }

  /**
   * Adds classes with the 'hover:' prefix.
   */
  hover(...classes: (string | null | undefined)[]): this {
    return this.withPrefix('hover:', ...classes);
  }

  /**
   * Adds classes with the 'focus:' prefix.
   */
  focus(...classes: (string | null | undefined)[]): this {
    return this.withPrefix('focus:', ...classes);
  }

  /**
   * Adds classes with the 'active:' prefix.
   */
  active(...classes: (string | null | undefined)[]): this {
    return this.withPrefix('active:', ...classes);
  }

  /**
   * Adds classes with the 'dark:' prefix for dark mode.
   */
  dark(...classes: (string | null | undefined)[]): this {
    return this.withPrefix('dark:', ...classes);
  }

  /**
   * Adds classes with the 'group-hover:' prefix.
   */
  groupHover(...classes: (string | null | undefined)[]): this {
    return this.withPrefix('group-hover:', ...classes);
  }

  /**
   * Adds classes with the 'focus-within:' prefix.
   */
  focusWithin(...classes: (string | null | undefined)[]): this {
    return this.withPrefix('focus-within:', ...classes);
  }

  /**
   * Adds responsive classes for the specified breakpoint.
   *
   * @param breakpoint Breakpoint prefix (sm, md, lg, xl, 2xl)
   * @param classes Classes to apply at this breakpoint
   */
  responsive(breakpoint: string, ...classes: (string | null | undefined)[]): this {
    return this.withPrefix(`${breakpoint}:`, ...classes);
  }

  /**
   * Adds classes for the 'sm' breakpoint (640px+).
   */
  sm(...classes: (string | null | undefined)[]): this {
    return this.responsive('sm', ...classes);
  }

  /**
   * Adds classes for the 'md' breakpoint (768px+).
   */
  md(...classes: (string | null | undefined)[]): this {
    return this.responsive('md', ...classes);
  }

  /**
   * Adds classes for the 'lg' breakpoint (1024px+).
   */
  lg(...classes: (string | null | undefined)[]): this {
    return this.responsive('lg', ...classes);
  }

  /**
   * Adds classes for the 'xl' breakpoint (1280px+).
   */
  xl(...classes: (string | null | undefined)[]): this {
    return this.responsive('xl', ...classes);
  }

  /**
   * Adds classes for the '2xl' breakpoint (1536px+).
   */
  xl2(...classes: (string | null | undefined)[]): this {
    return this.responsive('2xl', ...classes);
  }

  /**
   * Builds the final class string by joining all classes with spaces.
   */
  build(): string {
    return this.classes.join(' ');
  }

  /**
   * Returns the built class string.
   */
  toString(): string {
    return this.build();
  }
}

/**
 * Joins multiple class strings with spaces, filtering out null or empty values.
 *
 * @example
 * joinClasses("flex", "items-center", null, "gap-4")
 * // => "flex items-center gap-4"
 */
export function joinClasses(...classes: (string | null | undefined)[]): string {
  return classes.filter(c => c && c.trim()).join(' ');
}

/**
 * Conditionally returns a class string.
 *
 * @example
 * whenClass(isActive, "text-blue-600", "text-gray-600")
 */
export function whenClass(
  condition: boolean,
  className?: string,
  fallback?: string
): string | undefined {
  return condition ? className : fallback;
}

File: src/eQuantic.UI.Core/Styling/ClassBuilder.cs
#nullable enable

using System.Collections.Generic;
using System.Linq;

namespace eQuantic.UI.Core.Styling;

/// <summary>
/// Fluent builder for combining CSS classes.
/// Provides a clean, chainable API for building class strings.
/// </summary>
/// <example>
/// var classes = ClassBuilder.Create()
///     .Add("flex", "items-center")
///     .Add("gap-4", "p-6")
///     .When(isActive, "bg-blue-600", "text-white")
///     .Build();
/// </example>
public class ClassBuilder
{
    private readonly List<string> _classes = new();
    private readonly HashSet<string> _added = new();

    /// <summary>
    /// Creates a new empty ClassBuilder instance.
    /// </summary>
    public static ClassBuilder Create() => new();

    /// <summary>
    /// Creates a new ClassBuilder with initial classes.
    /// </summary>
    public static ClassBuilder Create(params string?[] initialClasses)
    {
        var builder = new ClassBuilder();
        builder.Add(initialClasses);
        return builder;
    }

    /// <summary>
    /// Adds one or more CSS classes.
    /// Duplicate classes are automatically ignored.
    /// </summary>
    public ClassBuilder Add(params string?[] classes)
    {
        foreach (var className in classes)
        {
            if (!string.IsNullOrWhiteSpace(className) && _added.Add(className))
            {
                _classes.Add(className);
            }
        }
        return this;
    }

    /// <summary>
    /// Conditionally adds classes if the condition is true.
    /// </summary>
    public ClassBuilder When(bool condition, params string?[] classes)
    {
        if (condition)
        {
            Add(classes);
        }
        return this;
    }

    /// <summary>
    /// Adds classes with a custom prefix.
    /// </summary>
    /// <example>
    /// .WithPrefix("hover:", "bg-gray-100", "scale-110") => "hover:bg-gray-100 hover:scale-110"
    /// .WithPrefix("dark:", "from-zinc-950 via-zinc-900") => "dark:from-zinc-950 dark:via-zinc-900"
    /// </example>
    public ClassBuilder WithPrefix(string prefix, params string?[] classes)
    {
        foreach (var className in classes)
        {
            if (!string.IsNullOrWhiteSpace(className))
            {
                // Split by whitespace to handle multi-class strings
                var individualClasses = className.Split(new[] { ' ', '\t', '\n', '\r' }, StringSplitOptions.RemoveEmptyEntries);
                foreach (var cls in individualClasses)
                {
                    Add($"{prefix}{cls}");
                }
            }
        }
        return this;
    }

    /// <summary>
    /// Adds classes with the 'hover:' prefix.
    /// </summary>
    public ClassBuilder Hover(params string?[] classes) => WithPrefix("hover:", classes);

    /// <summary>
    /// Adds classes with the 'focus:' prefix.
    /// </summary>
    public ClassBuilder Focus(params string?[] classes) => WithPrefix("focus:", classes);

    /// <summary>
    /// Adds classes with the 'active:' prefix.
    /// </summary>
    public ClassBuilder Active(params string?[] classes) => WithPrefix("active:", classes);

    /// <summary>
    /// Adds classes with the 'focus-within:' prefix.
    /// </summary>
    public ClassBuilder FocusWithin(params string?[] classes) => WithPrefix("focus-within:", classes);

    /// <summary>
    /// Adds classes with the 'placeholder:' prefix.
    /// </summary>
    public ClassBuilder Placeholder(params string?[] classes) => WithPrefix("placeholder:", classes);

    /// <summary>
    /// Adds classes with the 'visited:' prefix.
    /// </summary>
    public ClassBuilder Visited(params string?[] classes) => WithPrefix("visited:", classes);

    /// <summary>
    /// Adds classes with the 'disabled:' prefix.
    /// </summary>
    public ClassBuilder Disabled(params string?[] classes) => WithPrefix("disabled:", classes);

    /// <summary>
    /// Adds classes with the 'odd:' prefix.
    /// </summary>
    public ClassBuilder Odd(params string?[] classes) => WithPrefix("odd:", classes);

    /// <summary>
    /// Adds classes with the 'even:' prefix.
    /// </summary>
    public ClassBuilder Even(params string?[] classes) => WithPrefix("even:", classes);

    /// <summary>
    /// Adds classes with the 'first:' prefix.
    /// </summary>
    public ClassBuilder First(params string?[] classes) => WithPrefix("first:", classes);

    /// <summary>
    /// Adds classes with the 'last:' prefix.
    /// </summary>
    public ClassBuilder Last(params string?[] classes) => WithPrefix("last:", classes);

    /// <summary>
    /// Adds classes with the 'before:' prefix.
    /// </summary>
    public ClassBuilder Before(params string?[] classes) => WithPrefix("before:", classes);

    /// <summary>
    /// Adds classes with the 'after:' prefix.
    /// </summary>
    public ClassBuilder After(params string?[] classes) => WithPrefix("after:", classes);

    /// <summary>
    /// Adds classes with the 'group-hover:' prefix.
    /// </summary>
    public ClassBuilder GroupHover(params string?[] classes) => WithPrefix("group-hover:", classes);

    /// <summary>
    /// Adds classes with the 'dark:' prefix for dark mode.
    /// </summary>
    public ClassBuilder Dark(params string?[] classes) => WithPrefix("dark:", classes);

    /// <summary>
    /// Adds classes with the 'sm:' prefix.
    /// </summary>
    public ClassBuilder Sm(params string?[] classes) => WithPrefix("sm:", classes);

    /// <summary>
    /// Adds classes with the 'md:' prefix.
    /// </summary>
    public ClassBuilder Md(params string?[] classes) => WithPrefix("md:", classes);

    /// <summary>
    /// Adds classes with the 'lg:' prefix.
    /// </summary>
    public ClassBuilder Lg(params string?[] classes) => WithPrefix("lg:", classes);

    /// <summary>
    /// Adds classes with the 'xl:' prefix.
    /// </summary>
    public ClassBuilder Xl(params string?[] classes) => WithPrefix("xl:", classes);

    /// <summary>
    /// Adds classes with the '2xl:' prefix.
    /// </summary>
    public ClassBuilder Xl2(params string?[] classes) => WithPrefix("2xl:", classes);

    /// <summary>
    /// Builds the final class string by joining all classes with spaces.
    /// </summary>
    public string Build()
    {
        return string.Join(" ", _classes);
    }

    /// <summary>
    /// Implicit conversion to string.
    /// </summary>
    public static implicit operator string(ClassBuilder builder) => builder.Build();

    /// <summary>
    /// Returns the built class string.
    /// </summary>
    public override string ToString() => Build();
}

/// <summary>
/// Extension methods for ClassBuilder.
/// </summary>
public static class ClassBuilderExtensions
{
    /// <summary>
    /// Joins multiple class strings with spaces, filtering out null or empty values.
    /// </summary>
    public static string JoinClasses(params string?[] classes)
    {
        return string.Join(" ", classes.Where(c => !string.IsNullOrWhiteSpace(c)));
    }

    /// <summary>
    /// Conditionally returns a class string.
    /// </summary>
    public static string? WhenClass(bool condition, string? className, string? fallback = null)
    {
        return condition ? className : fallback;
    }
}
